In [ ]:
import pandas as pd

In [ ]:
url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing" # orderlines.csv
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
df_orderlines = pd.read_csv(path)

In [ ]:
url = "https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=drive_link" # products.csv
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
df_products = pd.read_csv(path)

In [ ]:
url = "https://drive.google.com/file/d/1Vu0q91qZw6lqhIqbjoXYvYAQTmVHh6uZ/view?usp=drive_link" # orders.csv
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
df_orders = pd.read_csv(path)

In [ ]:
url = "https://drive.google.com/file/d/1m1ThDDIYRTTii-rqM5SEQjJ8McidJskD/view?usp=drive_link" # brands.csv
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
df_brands = pd.read_csv(path)

First we create copies of orderlines, products and orders

In [ ]:
orders_df = df_orders.copy()

orderlines_df = df_orderlines.copy()

products_df = df_products.copy()

# 1. Duplicates
Then we check for duplicates in orders, orderlines and products

In [ ]:
products_df.duplicated().sum(), orders_df.duplicated().sum(), orderlines_df.duplicated().sum()

(np.int64(8746), np.int64(0), np.int64(0))

We clean product duplicates

In [ ]:
products_df = products_df.drop_duplicates().copy()
products_df.duplicated().sum(), orders_df.duplicated().sum(), orderlines_df.duplicated().sum()  ## Now we have no more duplicates

(np.int64(0), np.int64(0), np.int64(0))

In [ ]:
products_df.loc[products_df.duplicated(subset="sku", keep=False)]

sku                                               name  \
7992  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   
8000  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   

                                                   desc price promo_price  \
7992  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...  1729     1305.59   
8000  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...   NaN     1305.59   

      in_stock  type  
7992         0  1282  
8000         0  1282

In [ ]:
products_df.duplicated(subset="sku").sum()

np.int64(1)

There is one sku duplicate which we can also drop

In [ ]:
products_df = products_df.drop_duplicates(subset="sku", keep="first").copy()
products_df.duplicated(subset="sku").sum()

np.int64(0)

# 2. .info()

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


total_paid has 5 missing values since it is 226904 and others are 226909<br> created_date should become datetime datatype

In [ ]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


date should be a datetime datatype<br>
unit_price should be a float datatype

In [ ]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10579 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10579 non-null  object
 1   name         10579 non-null  object
 2   desc         10572 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10579 non-null  object
 5   in_stock     10579 non-null  int64 
 6   type         10529 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


price, promo_price & type as string formatted <br>
price, des and type with Null values

# 3. Missing values

**3.1 Orders**

In [ ]:
# total_paid has 5 missing values
orders_df['total_paid'].isna().value_counts(normalize=True)

total_paid
False    0.999978
True     0.000022
Name: proportion, dtype: float64

As there is such a tiny amount of missing values, we will simply delete these rows, as we have enough data without them.

In [ ]:
orders_df = orders_df.dropna(axis=0)
orders_df['total_paid'].isna().value_counts(normalize=True)

total_paid
False    1.0
Name: proportion, dtype: float64

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226904 non-null  int64  
 1   created_date  226904 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226904 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 8.7+ MB


**3.2 Orderlines**

There are no missing values in orderlines_df

**3.3 Products**

In [ ]:
products_df.isna().sum()

sku             0
name            0
desc            7
price          45
promo_price     0
in_stock        0
type           50
dtype: int64

In [ ]:
products_df["type"] = products_df["type"].fillna("missing")  ## Here we fill the missing 50 type with "missing"
products_df.loc[products_df["desc"].isna(), "desc"] = products_df.loc[products_df["desc"].isna(), "name"] ## desc will no longer has NaN in those rows — it contains the product name as a fallback.
products_df = products_df.dropna(subset="price").copy()
products_df.isna().sum()

sku            0
name           0
desc           0
price          0
promo_price    0
in_stock       0
type           0
dtype: int64

# 4. Datatypes

**4.1 Orders**

created_date should become datetime datatype

In [ ]:
orders_df["created_date"] = pd.to_datetime(orders_df["created_date"])
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226904 non-null  int64         
 1   created_date  226904 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226904 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 8.7+ MB


Then we add a month column to orders_df

In [ ]:
orders_df.loc[:,"month"] = orders_df["created_date"].dt.month
orders_df.head()

order_id        created_date  total_paid      state  month
0    241319 2017-01-02 13:35:40       44.99  Cancelled      1
1    241423 2017-11-06 13:10:02      136.15  Completed     11
2    242832 2017-12-31 17:40:03       15.76  Completed     12
3    243330 2017-02-16 10:59:38       84.98  Completed      2
4    243784 2017-11-24 13:35:19      157.86  Cancelled     11

Then we add another column with month name

In [ ]:
orders_df.loc[:,"month_name"] = orders_df["created_date"].dt.month_name()
orders_df.head()

order_id        created_date  total_paid      state  month month_name
0    241319 2017-01-02 13:35:40       44.99  Cancelled      1    January
1    241423 2017-11-06 13:10:02      136.15  Completed     11   November
2    242832 2017-12-31 17:40:03       15.76  Completed     12   December
3    243330 2017-02-16 10:59:38       84.98  Completed      2   February
4    243784 2017-11-24 13:35:19      157.86  Cancelled     11   November

Then we add another column with year

In [ ]:
orders_df.loc[:,"year"] = orders_df["created_date"].dt.year
orders_df.head()

order_id        created_date  total_paid      state  month month_name  year
0    241319 2017-01-02 13:35:40       44.99  Cancelled      1    January  2017
1    241423 2017-11-06 13:10:02      136.15  Completed     11   November  2017
2    242832 2017-12-31 17:40:03       15.76  Completed     12   December  2017
3    243330 2017-02-16 10:59:38       84.98  Completed      2   February  2017
4    243784 2017-11-24 13:35:19      157.86  Cancelled     11   November  2017

Then we add another column with day

In [ ]:
orders_df.loc[:,"day"] = orders_df["created_date"].dt.day
orders_df.head()

order_id        created_date  total_paid      state  month month_name  \
0    241319 2017-01-02 13:35:40       44.99  Cancelled      1    January   
1    241423 2017-11-06 13:10:02      136.15  Completed     11   November   
2    242832 2017-12-31 17:40:03       15.76  Completed     12   December   
3    243330 2017-02-16 10:59:38       84.98  Completed      2   February   
4    243784 2017-11-24 13:35:19      157.86  Cancelled     11   November   

   year  day  
0  2017    2  
1  2017    6  
2  2017   31  
3  2017   16  
4  2017   24

Finally we add another two columns with weekdays and day_of_week

In [ ]:
orders_df.loc[:,"weekday"] = orders_df["created_date"].dt.weekday
orders_df.loc[:,"day_of_week"] = orders_df["created_date"].dt.day_name()
orders_df.head()

order_id        created_date  total_paid      state  month month_name  \
0    241319 2017-01-02 13:35:40       44.99  Cancelled      1    January   
1    241423 2017-11-06 13:10:02      136.15  Completed     11   November   
2    242832 2017-12-31 17:40:03       15.76  Completed     12   December   
3    243330 2017-02-16 10:59:38       84.98  Completed      2   February   
4    243784 2017-11-24 13:35:19      157.86  Cancelled     11   November   

   year  day  weekday day_of_week  
0  2017    2        0      Monday  
1  2017    6        0      Monday  
2  2017   31        6      Sunday  
3  2017   16        3    Thursday  
4  2017   24        4      Friday

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226904 non-null  int64         
 1   created_date  226904 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226904 non-null  object        
 4   month         226904 non-null  int32         
 5   month_name    226904 non-null  object        
 6   year          226904 non-null  int32         
 7   day           226904 non-null  int32         
 8   weekday       226904 non-null  int32         
 9   day_of_week   226904 non-null  object        
dtypes: datetime64[ns](1), float64(1), int32(4), int64(1), object(3)
memory usage: 15.6+ MB


Need to convert month column (#4) to integer from object

In [ ]:
orders_df['month'] = pd.to_numeric(orders_df['month'], errors='coerce')
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226904 non-null  int64         
 1   created_date  226904 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226904 non-null  object        
 4   month         226904 non-null  int32         
 5   month_name    226904 non-null  object        
 6   year          226904 non-null  int32         
 7   day           226904 non-null  int32         
 8   weekday       226904 non-null  int32         
 9   day_of_week   226904 non-null  object        
dtypes: datetime64[ns](1), float64(1), int32(4), int64(1), object(3)
memory usage: 15.6+ MB


**4.2 Orderlines**

date should be a datetime datatype <br>
unit_price should be a float datatype but *first we have to deal with the 2 decimal points*

In [ ]:
orderlines_df["date"] = pd.to_datetime(orderlines_df["date"])
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  object        
 6   date              293983 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(2)
memory usage: 15.7+ MB


In [ ]:
# First we count the number of decimal points in the unit_price
orderlines_df['unit_price'].str.count(r"\.").value_counts()

unit_price
1    257814
2     36169
Name: count, dtype: int64

In [ ]:
# Count the rows with more than one `.`
mult_decimal_rows = (orderlines_df['unit_price'].str.count(r"\.")>1).sum()

# Find the percentage of corrupted rows
percent_corrupted = (100 * mult_decimal_rows / orderlines_df.shape[0])
print(f"{percent_corrupted:.2f}% of the rows in our DataFrame have multiple decimal points in the unit_price")

12.30% of the rows in our DataFrame have multiple decimal points in the unit_price


Now let us remove those decimal points

In [ ]:
# Boolean mask to find the orders that contain a price with multiple decimal points
multiple_decimal_mask = orderlines_df['unit_price'].str.count(r"\.") > 1

# Apply the boolean mask to the orderlines DataFrame. This way we can find the order_id of all the affected orders.
corrupted_order_ids = orderlines_df.loc[multiple_decimal_mask, "id_order"]

# Keep only the rows that do not have multiple decimal points
orderlines_df = orderlines_df.loc[~orderlines_df['id_order'].isin(corrupted_order_ids)]

In [ ]:
orderlines_df.shape[0]

216250

Now that all decimal points have been removed, we can convert the column unit_price to the correct datatype.

In [ ]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"], errors="coerce")
orderlines_df = orderlines_df.dropna().copy()
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 216250 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                216250 non-null  int64         
 1   id_order          216250 non-null  int64         
 2   product_id        216250 non-null  int64         
 3   product_quantity  216250 non-null  int64         
 4   sku               216250 non-null  object        
 5   unit_price        216250 non-null  float64       
 6   date              216250 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 13.2+ MB


The following lines are additional sanity checks and corrections to sort to inspect suspicious values, inspect a specific SKU for anomalies, make sure all prices are positive for revenue calculations.

In [ ]:
orderlines_df.sort_values(by="unit_price") # sku == APP1465
orderlines_df.query("sku == 'APP1465'")["unit_price"].value_counts()
orderlines_df["unit_price"] = orderlines_df["unit_price"].abs()

In [ ]:
orderlines_df.head()

id  id_order  product_id  product_quantity      sku  unit_price  \
0  1119109    299539           0                 1  OTT0133       18.99   
1  1119110    299540           0                 1  LGE0043      399.00   
2  1119111    299541           0                 1  PAR0071      474.05   
3  1119112    299542           0                 1  WDT0315       68.39   
4  1119113    299543           0                 1  JBL0104       23.74   

                 date  
0 2017-01-01 00:07:19  
1 2017-01-01 00:19:45  
2 2017-01-01 00:20:57  
3 2017-01-01 00:51:40  
4 2017-01-01 01:06:38

Now we drop the product_id column since it has no value

In [ ]:
#Remove product_id column
orderlines_df = orderlines_df.drop("product_id", axis = 1)

In [ ]:
orderlines_df

id  id_order  product_quantity      sku  unit_price  \
0       1119109    299539                 1  OTT0133       18.99   
1       1119110    299540                 1  LGE0043      399.00   
2       1119111    299541                 1  PAR0071      474.05   
3       1119112    299542                 1  WDT0315       68.39   
4       1119113    299543                 1  JBL0104       23.74   
...         ...       ...               ...      ...         ...   
293978  1650199    527398                 1  JBL0122       42.99   
293979  1650200    527399                 1  PAC0653      141.58   
293980  1650201    527400                 2  APP0698        9.99   
293981  1650202    527388                 1  BEZ0204       19.99   
293982  1650203    527401                 1  APP0927       13.99   

                      date  
0      2017-01-01 00:07:19  
1      2017-01-01 00:19:45  
2      2017-01-01 00:20:57  
3      2017-01-01 00:51:40  
4      2017-01-01 01:06:38  
...                    ...  
293978 2018-03-14 13:57:25  
293979 2018-03-14 13:57:34  
293980 2018-03-14 13:57:41  
293981 2018-03-14 13:58:01  
293982 2018-03-14 13:58:36  

[216250 rows x 6 columns]

In [ ]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 216250 entries, 0 to 293982
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                216250 non-null  int64         
 1   id_order          216250 non-null  int64         
 2   product_quantity  216250 non-null  int64         
 3   sku               216250 non-null  object        
 4   unit_price        216250 non-null  float64       
 5   date              216250 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(3), object(1)
memory usage: 11.5+ MB


In [ ]:
orderlines_df["unit_price"].describe()

count    216250.000000
mean        164.637261
std         219.629779
min           0.000000
25%          29.740000
50%          69.990000
75%         184.580000
max         999.990000
Name: unit_price, dtype: float64

**4.3 Products**

In [ ]:
products_df.head()

sku                                           name  \
0  RAI0007              Silver Rain Design mStand Support   
1  APP0023              Apple Mac Keyboard Keypad Spanish   
2  APP0025               Mighty Mouse Apple Mouse for Mac   
3  APP0072  Apple Dock to USB Cable iPhone and iPod white   
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   

   in_stock      type  
0         1      8696  
1         0  13855401  
2         0      1387  
3         0      1230  
4         1      1364

The column promo_price seems to be corrupted. Since the unit_price column in orderlines table can be used to evaluate discounts, we can delete promo_price. in_stock column is also not needed.

In [ ]:
#Remove in stock
products_df = products_df.drop("in_stock", axis=1)

In [ ]:
products_df


sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price promo_price  \
0           Aluminum support compatible with all MacBook  59.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                                 mouse Apple USB cable.     59     569.898   
3                  IPhone dock and USB Cable Apple iPod.     25     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
...                                                  ...    ...         ...   
19321  compact and portable stand vertically or horiz...  29.99     269.903   
19322  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19323  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19324  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19325  Backpack with capacity of 14 liter compartment...  69.95     649.903   

           type  
0          8696  
1      13855401  
2          1387  
3          1230  
4          1364  
...         ...  
19321     12282  
19322      1392  
19323      1392  
19324      1392  
19325      1392  

[10534 rows x 6 columns]

In [ ]:
#Remove promo_price
products_df = products_df.drop("promo_price", axis=1)

In [ ]:
products_df

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price      type  
0           Aluminum support compatible with all MacBook  59.99      8696  
1              USB ultrathin keyboard Apple Mac Spanish.     59  13855401  
2                                 mouse Apple USB cable.     59      1387  
3                  IPhone dock and USB Cable Apple iPod.     25      1230  
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99      1364  
...                                                  ...    ...       ...  
19321  compact and portable stand vertically or horiz...  29.99     12282  
19322  Backpack with capacity of 14 liter compartment...  69.95      1392  
19323  Backpack with capacity of 14 liter compartment...  69.95      1392  
19324  Backpack with capacity of 14 liter compartment...  69.95      1392  
19325  Backpack with capacity of 14 liter compartment...  69.95      1392  

[10534 rows x 5 columns]

In [ ]:
products_df.duplicated().sum()

np.int64(0)

In [ ]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10534 entries, 0 to 19325
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   sku     10534 non-null  object
 1   name    10534 non-null  object
 2   desc    10534 non-null  object
 3   price   10534 non-null  object
 4   type    10534 non-null  object
dtypes: object(5)
memory usage: 493.8+ KB


Now we remove duplicated dots and change data types float in column price

In [ ]:
# Boolean mask to find the orders that contain a price with multiple decimal points
multiple_decimal_mask_product = products_df['price'].str.count(r"\.") > 1

# Apply the boolean mask to the orderlines DataFrame.
corrupted_order_ids_product = products_df.loc[multiple_decimal_mask_product, "sku"]

# Keep only the rows that do not have multiple decimal points
products_df = products_df.loc[~products_df['sku'].isin(corrupted_order_ids_product)]

In [ ]:
products_df.shape

(10157, 5)

In [ ]:
products_df.loc[:, "price"] = pd.to_numeric(products_df["price"], errors="coerce")

In [ ]:
#Drop null values
products_df = products_df.dropna()

In [ ]:
products_df["price"].describe()

count    10157.000000
mean       663.124629
std       1355.237789
min          2.990000
25%         44.900000
50%        119.000000
75%        649.891000
max      15339.000000
Name: price, dtype: float64

In [ ]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 216250 entries, 0 to 293982
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                216250 non-null  int64         
 1   id_order          216250 non-null  int64         
 2   product_quantity  216250 non-null  int64         
 3   sku               216250 non-null  object        
 4   unit_price        216250 non-null  float64       
 5   date              216250 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(3), object(1)
memory usage: 11.5+ MB
